File: sdks/ruby/lib/blok/testing/mock_context.rb
# frozen_string_literal: true

module Blok
  module Testing
    # MockContext provides a fluent builder for creating test workflow contexts.
    #
    # @example
    #   ctx = Blok::Testing::MockContext.new
    #     .with_id("test-123")
    #     .with_body({ "name" => "World" })
    #     .with_var("key", "value")
    #     .build
    #
    class MockContext
      def initialize
        @id            = "test-execution-id"
        @workflow_name = "test-workflow"
        @workflow_path = "/workflows/test"
        @body          = {}
        @headers       = {}
        @method        = "POST"
        @url           = "/test"
        @vars          = {}
        @env           = {}
      end

      # Set the execution ID.
      # @param id [String]
      # @return [self]
      def with_id(id)
        @id = id
        self
      end

      # Set the workflow name and path.
      # @param name [String]
      # @param path [String]
      # @return [self]
      def with_workflow(name, path = "/workflows/#{name}")
        @workflow_name = name
        @workflow_path = path
        self
      end

      # Set the request body.
      # @param body [Hash] Request body
      # @return [self]
      def with_body(body)
        @body = body
        self
      end

      # Set the request headers.
      # @param headers [Hash<String, String>]
      # @return [self]
      def with_headers(headers)
        @headers = headers
        self
      end

      # Set the request method.
      # @param method [String] HTTP method
      # @return [self]
      def with_method(method)
        @method = method
        self
      end

      # Set a context variable.
      # @param key [String] Variable name
      # @param value [Object] Variable value
      # @return [self]
      def with_var(key, value)
        @vars[key] = value
        self
      end

      # Set an environment variable.
      # @param key [String] Variable name
      # @param value [String] Variable value
      # @return [self]
      def with_env(key, value)
        @env[key] = value
        self
      end

      # Build the Context instance.
      # @return [Blok::Types::Context]
      def build
        request = Types::Request.new(
          body:     @body,
          headers:  @headers,
          method:   @method,
          url:      @url,
          base_url: "http://localhost:8080"
        )

        Types::Context.new(
          id:            @id,
          workflow_name: @workflow_name,
          workflow_path: @workflow_path,
          request:       request,
          vars:          @vars,
          env:           @env
        )
      end
    end
  end
end

File: sdks/php/src/Testing/MockContext.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Testing;

use Blok\Blok\Types\Context;
use Blok\Blok\Types\Request;
use Blok\Blok\Types\Response;

/**
 * MockContext provides a fluent builder for creating test Context instances.
 */
final class MockContext
{
    private string $id = 'test-execution-id';
    private string $workflowName = 'test-workflow';
    private string $workflowPath = '/workflows/test';
    private mixed $body = [];
    /** @var array<string, string> */
    private array $headers = [];
    private string $method = 'POST';
    private string $url = '/test';
    /** @var array<string, mixed> */
    private array $vars = [];
    /** @var array<string, string> */
    private array $env = [];
    /** @var array<string, string> */
    private array $params = [];
    /** @var array<string, string> */
    private array $query = [];

    /**
     * Create a new MockContext builder.
     */
    public static function create(): self
    {
        return new self();
    }

    /**
     * Set the execution ID.
     */
    public function withId(string $id): self
    {
        $this->id = $id;
        return $this;
    }

    /**
     * Set the workflow name and path.
     */
    public function withWorkflow(string $name, string $path): self
    {
        $this->workflowName = $name;
        $this->workflowPath = $path;
        return $this;
    }

    /**
     * Set the request body.
     */
    public function withBody(mixed $body): self
    {
        $this->body = $body;
        return $this;
    }

    /**
     * Set the request headers.
     *
     * @param array<string, string> $headers
     */
    public function withHeaders(array $headers): self
    {
        $this->headers = $headers;
        return $this;
    }

    /**
     * Set the request method.
     */
    public function withMethod(string $method): self
    {
        $this->method = $method;
        return $this;
    }

    /**
     * Set the request URL.
     */
    public function withUrl(string $url): self
    {
        $this->url = $url;
        return $this;
    }

    /**
     * Set a context variable.
     */
    public function withVar(string $key, mixed $value): self
    {
        $this->vars[$key] = $value;
        return $this;
    }

    /**
     * Set an environment variable.
     */
    public function withEnv(string $key, string $value): self
    {
        $this->env[$key] = $value;
        return $this;
    }

    /**
     * Set request params.
     *
     * @param array<string, string> $params
     */
    public function withParams(array $params): self
    {
        $this->params = $params;
        return $this;
    }

    /**
     * Set request query parameters.
     *
     * @param array<string, string> $query
     */
    public function withQuery(array $query): self
    {
        $this->query = $query;
        return $this;
    }

    /**
     * Build the Context instance.
     */
    public function build(): Context
    {
        return new Context(
            id: $this->id,
            workflowName: $this->workflowName,
            workflowPath: $this->workflowPath,
            request: new Request(
                body: $this->body,
                headers: $this->headers,
                params: $this->params,
                query: $this->query,
                method: $this->method,
                url: $this->url,
                baseUrl: 'http://localhost:8080',
            ),
            response: new Response(),
            vars: $this->vars,
            env: $this->env,
        );
    }
}
